cpp/inc/bond/core/maybe.h

```diff
@@ -14,143 +18,249 @@ namespace bond
-        : _value(std::move(that._value)),
-          _nothing(std::move(that._nothing))
+    /// @brief Check if this object contains nothing
+    bool is_nothing() const
     {
-        that._nothing = true;
+        return static_cast<bool>(_value);
     }
 
-    template <typename Allocator>
-    maybe(maybe&& that, const Allocator& alloc)
-        : _value(std::move(that._value), alloc),
-          _nothing(that._nothing)
+    /// @brief Set to nothing
+    void set_nothing()
     {
-        that._nothing = true;
+        _value = boost::none;
     }
 
-    maybe(const maybe& that) = default;
-    maybe& operator=(const maybe& that) = default;
-
-    /// @brief Constructor from a value T
-    explicit
-    maybe(const T& value)
-        : _value(value),
-          _nothing(false)
+    /// @brief Construct in place
+    ///
+    /// @since 8.0.0
+    template<class... Args>
+    void emplace(Args&&... args)
     {
+        _value.emplace(std::forward<Args>(args)...);
     }
 
-    template <typename Allocator>
-    explicit maybe(const Allocator& alloc)
-        : _value(alloc),
-          _nothing(true)
-    {}
+    /// @brief Reference to the value
+    /// @throw NothingException if the object contains nothing
+    T& value()
+    {
+        if (is_nothing())
+        {
+            NothingException();
+        }
 
-    template <typename Allocator>
-    maybe(const maybe& that, const Allocator& alloc)
-        : _value(that._value, alloc),
-          _nothing(that._nothing)
-    {}
+        return _value.get();
+    }
 
-    /// @brief Swap this object with that object
-    void swap(maybe& that)
+    /// @brief Constant reference to the value
+    /// @throw NothingException if the object contains nothing
+    const T& value() const
     {
-        using std::swap;
+        if (is_nothing())
+        {
+            NothingException();
+        }
 
-        swap(_nothing, that._nothing);
-        swap(_value, that._value);
+        return _value.get();
     }
 
-    /// @brief Check if this object contains nothing
-    bool is_nothing() const
+    /// @brief Convert to constant reference to value
+    /// @throw NothingException if the object contains nothing
+    operator const T&() const
     {
-        return _nothing;
+        return value();
     }
 
-    /// @brief Set to default instance of T and return reference to that value
-    T& set_value()
+protected:
+    maybe_common() = default;
+    maybe_common(const maybe_common&) = default;
+    maybe_common& operator=(const maybe_common&) = default;
+
+    explicit maybe_common(const T& value)
+        : _value(value)
+    { }
+
+    template <typename Alloc>
+    maybe_common(const T& value, const Alloc& alloc)
+        : _value(boost::in_place_init, value, alloc)
+    { }
+
+    maybe_common(maybe_common&& that) BOND_NOEXCEPT_IF(std::is_nothrow_move_constructible<boost::optional<T>>::value)
+        : _value(that._value)
```

Comment:
Missing `std::move`.

Reply:
I mistakenly thought `move` wasn't needed because `that` is `maybe_common&&`. However, `that` is  an lvalue (of an rvalue ref). Will switch.